Sources/AsyncAlgorithms/AsyncWithLatestFrom2Sequence.swift

```diff
@@ -0,0 +1,261 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public extension AsyncSequence {
+  /// Combines `self` with two other ``AsyncSequence`` into a single ``AsyncSequence`` where each
+  /// element from `self` is aggregated to the latest known elements from the `other` sequences (if any) as a tuple.
+  ///
+  /// Remark: as the `other` sequences are being iterated over in the context of their own ``Task``, there is no guarantee
+  /// that their latest know elements are the ones that have just been produced when the base sequence produces its next element.
+  ///
+  /// ```
+  /// let base = AsyncChannel<Int>()
+  /// let other1 = AsyncChannel<String>()
+  /// let other2 = AsyncChannel<String>()
+  /// let sequence = base.withLatest(from: other1, other2)
+  ///
+  /// Task {
+  ///   for element in await sequence {
+  ///    print(element)
+  ///   }
+  /// }
+  ///
+  /// await other1.send("a")
+  /// await other1.send("b")
+  ///
+  /// await other2.send("c")
+  /// await other2.send("d")
+  ///
+  /// ... later in the application flow
+  ///
+  /// await base.send(1)
+  ///
+  /// // will print: (1, "b", "d")
+  /// ```
+  ///
+  /// - Parameters:
+  ///   - other1: the first other ``AsyncSequence``
+  ///   - other2: the second other ``AsyncSequence``
+  /// - Returns: an ``AsyncWithLatestFrom2Sequence`` where elements are a tuple of an element from `self` and the
+  /// latest known elements (if any) from the `other` sequences.
+  func withLatest<Other1: AsyncSequence, Other2: AsyncSequence>(
+    from other1: Other1,
+    _ other2: Other2
+  ) -> AsyncWithLatestFrom2Sequence<Self, Other1, Other2> {
+    AsyncWithLatestFrom2Sequence(self, other1, other2)
+  }
+}
+
+/// ``AsyncWithLatestFrom2Sequence`` is an ``AsyncSequence`` where elements are a tuple of an element from `base` and the
+/// latest known element (if any) from the `other` sequences.
+public struct AsyncWithLatestFrom2Sequence<Base: AsyncSequence, Other1: AsyncSequence, Other2: AsyncSequence>: AsyncSequence {
+  public typealias Element = (Base.Element, Other1.Element, Other2.Element)
+  public typealias AsyncIterator = Iterator
+
+  let base: Base
+  let other1: Other1
+  let other2: Other2
+
+  init(_ base: Base, _ other1: Other1, _ other2: Other2) {
+    self.base = base
+    self.other1 = other1
+    self.other2 = other2
+  }
+
+  public func makeAsyncIterator() -> Iterator {
```

Comment:
We should make sure that only a single Iterator can be created. The problem with multiple iterators is that it leads to non-deterministic behaviour where elements are ending up in.

Reply:
Can you be a bit more specific, I don't understand what you are asking for ?

Reply:
Sure, the problem here is a bit broader to the `AsyncSequence` ecosystem and I am planning to write something up soon. Right now in all the `AsyncSequence` implementations the `makeAsyncIterator` method can be called multiple times. Each of these iterators is almost always using the same underlying storage. So the actual implementation is that the algorithms are unicast `AsyncSequence`s however since they allow multiple iterators are getting into an awkward state where multiple `Task`s can consume them now. This causes all kinds of issues like non-deterministic behaviour who gets an element. Furthermore, cancellation handling is becoming problematic since if two `Task`s are consuming the same `AsyncSequence` and one gets cancelled it almost always cancels the other consumer as well. I currently consider it best practice to just guard that only a single Iterator can be created and `fatalError` on subsequent calls.

The current behaviour is fine for `Sequence`s since the they don't have this time aspect to them and are most of the time referencing an underlying content. For our `AsyncSequence`s the content is almost never buffered but consumed on the time stream.

As I said, I am planning to write more up on this soonish just wanted to make sure to put this out here already.

Reply:
Thanks for the detailed explanation. I'm not sure I see a problem since a call to `makeAsyncIterator()` creates a new iterator that has its own properties and storage. Nothing is shared between several iterators.

Am I missing something ?

For AsyncSequences that have a storage shared across consumers (like AsyncChannel or an hypothetical AsyncShareSequence),  the locking mechanism enforces the good behaviour right ?

Reply:
Ah sorry I missed that the whole core implementation is inside the Iterator. While this makes things a bit better since the state doesn't interfere it makes still for some weird behaviour IMO:

```swift
let other = AsyncChannel<>
let base = AsyncStream<>

let sequence = base.withLatest(from: other)

Task {
    for await element in sequence {}
}


Task {
    for await element in sequence {}
}
```

You cannot tell from this code what is going to happen since the way all of the sequences involved in this little snippet right now allow creating multiple iterators. While your implementation is multi-cast all the other ones are not. IMO, what I personally think we should do is implement almost all our algorithms and base sequences as unicast (unless they have different semantics) and provide `share/multicast` algorithms. This makes reasoning about the code in the end way easier

```swift
let other = AsyncChannel<>
let base = AsyncStream<>

let sequence = base.shared().withLatest(from: other.shared())
```

I don't think you have to change your PR here right away since this is still in discussion and like I said I wanna open up a broader discussion around this soonish. Just wanted to share more details with you in why I think we need to be careful around these things.

Reply:
Ok I understand. This is something I struggled with also. But it is mainly due to the fact that although AsyncChannel and AsyncStream are compatible with a multi producer/multi consumer paradigm, their output is not shared and the elements will be distributed across the consumers.

I guess this is what channels are for. If I refer to the Kotlin implementation it does exactly that. 

I've open sourced a project `AsyncExtensions` before `swift-async-algorithms` was announced :-) and I'm trying to address notably the share operator in there, as well as shared "channels". Now that `swift-async-algorithms` has been announced, I'm rethinking my repo as a companion to `swift-async-algorithms`. It means I will add things that don't yet exist in `swift-async-algorithms` and try to suggest them to this main repo step by step.

The repo is a work in progress (https://github.com/sideeffect-io/AsyncExtensions/tree/refactor) and the README is not up-to-date but there are some interesting operators in there, that were challenging, like `.switchToLatest()`.

Reply:
I think the `AsyncChannel` being MPMC makes sense; however, I am still very torn if `AsyncStream` should be like that. `AsyncChannel` does handle cancellation better since it only removes the continuations for that specific `Task` whereas `AsyncStream` cancels the whole thing which makes it very hard to use safely.

But we are learning all of this along the way so we should just come up with some rules around what to expect!

W.r.t. your package. I really like it! I also have a rudimentary implementation of `share` lying around; however, as @phausler said we first need to stabilise the algos that are in here and then add more.

Reply:
BTW, speaking of AsyncChannel cancellation, I have on ongoing PR regarding cancellation when several producers in different tasks -> https://github.com/apple/swift-async-algorithms/pull/184